### PR TITLE
[SCB-925] Springmvc, when have defaultValue, required should be false

### PR DESCRIPTION
**Pull request by weichao666**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by weichao666** [comment]
Springmvc, 
Generating schema：
1. if do not write defaultValue， required's value is what user write
2. if have written defaultValue，No matter how the user writes，required is false

running：
1. When param is not transmitted，if required is true, report error
2. When param is not transmitted , if required is false, handle it with defaultValue

File: integration-tests/springmvc-tests/springmvc-tests-common/src/test/java/org/apache/servicecomb/demo/springmvc/tests/endpoints/CodeFirstSpringmvcSimplifiedMappingAnnotation.java [hunk 1]
```diff
@@ -101,7 +101,7 @@ public String fileUpload(@RequestPart(name = "file1") MultipartFile file1,
 
   @PostMapping(path = "/uploadWithoutAnnotation", produces = MediaType.TEXT_PLAIN_VALUE)
   public String fileUploadWithoutAnnotation(MultipartFile file1, MultipartFile file2,
-      @RequestAttribute("name") String name) {
+      @RequestAttribute(name = "name", required = false) String name) {
```
Comment (liubao68) [inline integration-tests/springmvc-tests/springmvc-tests-common/src/test/java/org/apache/servicecomb/demo/springmvc/tests/endpoints/CodeFirstSpringmvcSimplifiedMappingAnnotation.java]: Do we have test cases that name is null?

File: integration-tests/springmvc-tests/springmvc-tests-common/src/test/java/org/apache/servicecomb/demo/springmvc/tests/endpoints/CodeFirstSpringmvc.java [hunk 2]
```diff
@@ -98,7 +98,7 @@ public String fileUpload(@RequestPart(name = "file1") MultipartFile file1,
 
   @RequestMapping(path = "/uploadWithoutAnnotation", method = RequestMethod.POST, produces = MediaType.TEXT_PLAIN_VALUE, consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
   public String fileUploadWithoutAnnotation(MultipartFile file1, MultipartFile file2,
-      @RequestAttribute("name") String name) {
+      @RequestAttribute(name = "name", required = false) String name) {
```
Comment (liubao68) [inline integration-tests/springmvc-tests/springmvc-tests-common/src/test/java/org/apache/servicecomb/demo/springmvc/tests/endpoints/CodeFirstSpringmvc.java]: Do we have test cases that name is null?

**Comment by liubao68** [comment]
This PR is a behavior change, need to notice in RN

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/CookieProcessorCreator.java [hunk 3]
```diff
@@ -51,6 +53,9 @@ public Object getValue(HttpServletRequest request) throws Exception {
         if (paramPath.equals(cookie.getName())) {
           value = cookie.getValue();
           if (value == null) {
+            if (isRequired()) {
```
Comment (wujimin) [inline common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/CookieProcessorCreator.java]: line 48, missed required and default value logic

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/FormProcessorCreator.java [hunk 4]
```diff
@@ -55,6 +57,9 @@ public Object getValue(HttpServletRequest request) {
 
       Object value = request.getParameter(paramPath);
       if (value == null) {
```
Comment (wujimin) [inline common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/FormProcessorCreator.java]: line 51/55, missed required and default value logic

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/HeaderProcessorCreator.java [hunk 5]
```diff
@@ -57,6 +59,9 @@ public Object getValue(HttpServletRequest request) throws Exception {
       } else {
         value = request.getHeader(paramPath);
         if (value == null) {
+          if (isRequired()) {
```
Comment (wujimin) [inline common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/HeaderProcessorCreator.java]: line 55, missed required and default value logic

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/PathProcessorCreator.java [hunk 6]
```diff
@@ -36,8 +36,8 @@
   public static final String PARAMTYPE = "path";
 
   public static class PathProcessor extends AbstractParamProcessor {
-    public PathProcessor(String paramPath, JavaType targetType, Object defaultValue) {
-      super(paramPath, targetType, defaultValue);
+    public PathProcessor(String paramPath, boolean required, JavaType targetType, Object defaultValue) {
+      super(paramPath, required, targetType, defaultValue);
```
Comment (wujimin) [inline common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/PathProcessorCreator.java]: path parameter must be required, can not be null

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/QueryProcessorCreator.java [hunk 7]
```diff
@@ -60,6 +62,9 @@ public Object getValue(HttpServletRequest request) throws Exception {
           }
         }
         if (value == null) {
+          if (isRequired()) {
```
Comment (wujimin) [inline common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/QueryProcessorCreator.java]: line 53, missed required and default value logic

File: swagger/swagger-generator/generator-springmvc/src/main/java/org/apache/servicecomb/swagger/generator/springmvc/processor/annotation/CookieValueAnnotationProcessor.java [hunk 8]
```diff
@@ -40,6 +41,11 @@ protected void fillParameter(Object annotation, OperationGenerator operationGene
       CookieParameter parameter) {
     super.fillParameter(annotation, operationGenerator, paramIdx, parameter);
 
+    String defaultValue = getAnnotationParameterDefaultValue(annotation);
```
Comment (wujimin) [inline swagger/swagger-generator/generator-springmvc/src/main/java/org/apache/servicecomb/swagger/generator/springmvc/processor/annotation/CookieValueAnnotationProcessor.java]: just get it from "parameter"?

File: swagger/swagger-generator/generator-springmvc/src/main/java/org/apache/servicecomb/swagger/generator/springmvc/processor/annotation/PathVariableAnnotationProcessor.java [hunk 9]
```diff
@@ -39,6 +40,11 @@ protected void fillParameter(Object annotation, OperationGenerator operationGene
       PathParameter parameter) {
     super.fillParameter(annotation, operationGenerator, paramIdx, parameter);
 
+    String defaultValue = getAnnotationParameterDefaultValue(annotation);
```
Comment (wujimin) [inline swagger/swagger-generator/generator-springmvc/src/main/java/org/apache/servicecomb/swagger/generator/springmvc/processor/annotation/PathVariableAnnotationProcessor.java]: path required always be true, no default definition

File: swagger/swagger-generator/generator-springmvc/src/main/java/org/apache/servicecomb/swagger/generator/springmvc/processor/annotation/RequestAttributeAnnotationProcessor.java [hunk 10]
```diff
@@ -39,6 +40,11 @@ protected void fillParameter(Object annotation, OperationGenerator operationGene
       FormParameter parameter) {
     super.fillParameter(annotation, operationGenerator, paramIdx, parameter);
 
+    String defaultValue = getAnnotationParameterDefaultValue(annotation);
```
Comment (wujimin) [inline swagger/swagger-generator/generator-springmvc/src/main/java/org/apache/servicecomb/swagger/generator/springmvc/processor/annotation/RequestAttributeAnnotationProcessor.java]: just get it from "parameter"?

File: swagger/swagger-generator/generator-springmvc/src/main/java/org/apache/servicecomb/swagger/generator/springmvc/processor/annotation/RequestHeaderAnnotationProcessor.java [hunk 11]
```diff
@@ -40,6 +41,11 @@ protected void fillParameter(Object annotation, OperationGenerator operationGene
       HeaderParameter parameter) {
     super.fillParameter(annotation, operationGenerator, paramIdx, parameter);
 
+    String defaultValue = getAnnotationParameterDefaultValue(annotation);
```
Comment (wujimin) [inline swagger/swagger-generator/generator-springmvc/src/main/java/org/apache/servicecomb/swagger/generator/springmvc/processor/annotation/RequestHeaderAnnotationProcessor.java]: just get it from "parameter"?

File: swagger/swagger-generator/generator-springmvc/src/main/java/org/apache/servicecomb/swagger/generator/springmvc/processor/annotation/RequestParamAnnotationProcessor.java [hunk 12]
```diff
@@ -40,6 +41,11 @@ protected void fillParameter(Object annotation, OperationGenerator operationGene
       QueryParameter parameter) {
     super.fillParameter(annotation, operationGenerator, paramIdx, parameter);
 
+    String defaultValue = getAnnotationParameterDefaultValue(annotation);
```
Comment (wujimin) [inline swagger/swagger-generator/generator-springmvc/src/main/java/org/apache/servicecomb/swagger/generator/springmvc/processor/annotation/RequestParamAnnotationProcessor.java]: just get it from "parameter"?

File: swagger/swagger-generator/generator-springmvc/src/main/java/org/apache/servicecomb/swagger/generator/springmvc/processor/annotation/RequestPartAnnotationProcessor.java [hunk 13]
```diff
@@ -39,6 +40,11 @@ protected void fillParameter(Object annotation, OperationGenerator operationGene
       FormParameter parameter) {
     super.fillParameter(annotation, operationGenerator, paramIdx, parameter);
 
+    String defaultValue = getAnnotationParameterDefaultValue(annotation);
```
Comment (wujimin) [inline swagger/swagger-generator/generator-springmvc/src/main/java/org/apache/servicecomb/swagger/generator/springmvc/processor/annotation/RequestPartAnnotationProcessor.java]: just get it from "parameter"?

**Comment by wujimin** [comment]
so many code need to be modified, maybe we can create multiple issues for the reviews?

**Comment by weichao666** [comment]
after this pr, integration-tests has been modified, i can't handle these conflicts well, so i'll close this pr and rewrite this function based on the latest code.

**Pull request closed without merging**